### ISSUE TITLE
Serifs cut at the edge of the view while in CRE

### ISSUE THREAD

**Issue by ptrm** [koreader/koreader#4577]
* KOReader version: v2019.02-19-g0113708_2019-02-08, but is present for over a year AFAIR
* Device: Kobos (Aura One, Clara, Forma), Android 9

#### Issue
Some serifs at the edge of a screen get cut by the margin:
![20190210_164548](https://user-images.githubusercontent.com/366258/52535846-99f4bf00-2d53-11e9-90c3-f4e3ce7afce1.jpg)

#### Steps to reproduce
Use Noto Serif font, issue is also present for [Baskervaldx](https://ctan.org/pkg/baskervaldx) and [Libertinus Serif 6.6](https://github.com/libertinus-fonts/libertinus) as well as other serif fonts. It should be visible for any epub which happens to have an 'f' at the end of line ;) It was present for all hinting, kerning and word gap settings (haven't tried all combinations, though ;) )

I was convinced it was a known issue, but no such  issue popped in search resultsl for my queries.

**Comment by poire-z** [comment]
I never noticed that, but now that you point it out, I indeed see it... with many fonts.
Could it be that fonts lie about the length of the `f` glyph to somehow bring nearer the next glyph and make some kind of ligature?

**Comment by Frenzie** [comment]
I can confirm that it happens in the book I was just reading (my first ever fb2, half as an experiment — beautiful footnotes btw, @poire-z ;-) ). But it only seems to happen on the f, not on serifs in general.

**Comment by poire-z** [comment]
(These footnotes in fb2 were already there in crengine - I just brought these cuties to epub :)

**Comment by ptrm** [comment]
> But it only seems to happen on the f, not on serifs in general.

True, the issue title is too general ;)

It is also visible in selections:
![20190210_171603](https://user-images.githubusercontent.com/366258/52536215-e5a96780-2d57-11e9-8011-1c32af095c84.jpg)

**Comment by poire-z** [comment]
If you can build crengine, can you check how it does with 2 lines commented out (don't trust the line numbers of the diff, just find the `getAdditionalCharWidth()` function)
```diff
--- a/crengine/src/lvtextfm.cpp
+++ b/crengine/src/lvtextfm.cpp
@@ -1039,23 +1039,24 @@ public:
     /// checks whether to add more space after italic character
     int getAdditionalCharWidth( int pos, int maxpos ) {
         if (m_text[pos]==0) // object
             return 0; // no additional space
         LVFont * font = (LVFont*)m_srcs[pos]->t.font;
         if (!font)
             return 0; // no font
-        if ( !font->getItalic() )
-            return 0; // not italic
+//        if ( !font->getItalic() )
+//            return 0; // not italic
         if ( pos<maxpos-1 && m_srcs[pos+1]==m_srcs[pos] )
             return 0; // the same font, non-last char
         // need to measure
         LVFont::glyph_info_t glyph;
         if ( !font->getGlyphInfo(m_text[pos], &glyph, '?') )
             return 0;
         int delta = glyph.originX - glyph.width + glyph.blackBoxX;//fix the italic CJK character
+        if (delta > 0) printf("delta: %d\n", delta);
         return delta > 0 ? delta : 0;
     }
```
With a sufficiently large font size, I seem to get 1px added for each `of`, may be some others 1px for `d` or `n` - but I'm not sure if that's enough to show the full `f` - better eyes than mine should tell if it is :)

**Comment by Frenzie** [comment]
@poire-z You don't even need good eyes, just big letters. :-P

Your suggested change works well for this specific case.

![screenshot_2019-02-10_17-30-39](https://user-images.githubusercontent.com/202757/52536362-c4e21180-2d59-11e9-9cde-1736dc6c8621.png)

![screenshot_2019-02-10_17-29-57](https://user-images.githubusercontent.com/202757/52536365-c9a6c580-2d59-11e9-93a6-9b33616a9e11.png)

**Comment by NiLuJe** [comment]
I'm not quite sure what's at play here, but there's definitely what appears to be some aggressive kerning at play here, I can confirm the underlying cause w/ FBink too.

![fl](https://user-images.githubusercontent.com/111974/52537307-83effa00-2d65-11e9-9f15-e5d5754caf49.png)

```
fbink -t regular=/mnt/onboard/.adds/koreader/fonts/noto/NotoSerif-Regular.ttf,size=48 -Mm "fl" -vc

[FBInk] Enabled Kobo Mark 7 quirks
[FBInk] Clock tick frequency appears to be 100 Hz
[FBInk] Screen density set to 300 dpi
[FBInk] Variable fb info: 1440x1920, 32bpp @ rotation: 3 (Counter Clockwise, 270°)
[FBInk] Fontsize set to 32x32 (IBM base glyph size: 8x8)
[FBInk] Line length: 45 cols, Page size: 60 rows
[FBInk] Horizontal fit is perfect!
[FBInk] Fixed fb info: ID is "mxc_epdc_fb", length of fb mem: 11796480 bytes & line length: 5760 bytes
[FBInk] Pen colors set to #000000 for the foreground and #FFFFFF for the background
Loading font '/mnt/onboard/.adds/koreader/fonts/noto/NotoSerif-Regular.ttf' for the Regular style
Initialized libunibreak
[FBInk] Font '/mnt/onboard/.adds/koreader/fonts/noto/NotoSerif-Regular.ttf' loaded
Printing string 'fl' @ 48pt, honoring the following margins { Top: 0px, Bottom: 0px, Left: 0px, Right: 0px } (formatted: N, overlay: N, no BG: N, no FG: N, inverted: N, flashing: N, centered: Y, H align: 0, halfway: Y, V align: 0, clear screen: Y, dithered: N, skip refresh: N)
Printing OpenType text.
Unformatted text defaulting to Regular font style
Max BL: 158  Max Desc: -42  Max LG: 0
Finished looking for linebreaks
Current Measured LW: 65  Line# 0
Current Measured LW: 98  Line# 0
1 lines to be printed
Maximum printable height is 1920
Actual print height is 200
Max LW: 1440  Max LH: 200  Max BL: 158  FntSize: 200
Finished printing line# 0
Refreshing region from LEFT: 671, TOP: 860, WIDTH: 98, HEIGHT: 200
```

**Comment by NiLuJe** [comment]
Relatively similar rendering w/ GIMP, too.

**Comment by ptrm** [comment]
That's what the letter boundary seems to be defined like. In comparison with capital F:
![smartselect_20190210-190751_bvnc free](https://user-images.githubusercontent.com/366258/52537451-54da8800-2d67-11e9-9ce1-20a3296f8de5.jpg)
![smartselect_20190210-190809_bvnc free](https://user-images.githubusercontent.com/366258/52537453-5a37d280-2d67-11e9-8a1a-c98a2c509fd0.jpg)

**Comment by poire-z** [comment]
Regarding my above quick patch (simply doing that getAdditionalCharWidth() for the last char on the line everytime and not only when it's in an italic font), there's a quite huge performance drop...
My regular performance test with a big book, that always load in 13s and render in 5s, would still load in 13s, but render in 13s too...
So, we'll have to be more clever, like use another cache for these glyph metrics... or more dirty and just do it for the letter `f` :)

**Comment by ptrm** [comment]
> So, we'll have to be more clever, like use another cache for these glyph metrics... or more dirty and just do it for the letter `f` :)

Looking at the designs, haven't tested it though, italics might provode other cases, like "g", for example:
![smartselect_20190210-194204_bvnc free](https://user-images.githubusercontent.com/366258/52537903-3c20a100-2d6c-11e9-8709-c73095029d7f.jpg)

Major italic "T" also seems to lean out of it bounds.

**Comment by poire-z** [comment]
The original bit of code I modified is there to currently just deal with italic, so your italic letters should display fine - and I guess there are many such italic letters, so it was noticed easily and that `getAdditionalCharWidth()` added just for them.
We just discovered that there are actually non-italic letters that should benefit from this additional space :)

**Comment by poire-z** [comment]
There is also already something to deal with space on the left, as on your italic `g`:
```
    /// checks whether to add more space on left before italic character
    int getAdditionalCharWidthOnLeft( int pos ) {
```

**Comment by NiLuJe** [comment]
Yeah, I don't see a better solution other than a cache of the "effective" metrics (and it'll have to depend on the font-size, too) :/.

**Comment by NiLuJe** [comment]
Although, taking an actual llok at the font, that's the LSB & RSB, the glyph metrics should provide 'em. They just happen to be negative in these cases, which the code might not be expecting?

**Comment by ptrm** [comment]
> The original bit of code I modified is there to currently just deal with italic

True... Overlooked this ;) Then there are fonts (i'm looking for edge cases now ;) ), like calligraphy ones (not that I would use them on an ereader ;) ), which have more characters easily cut:
![smartselect_20190210-200205_koreader](https://user-images.githubusercontent.com/366258/52538128-21036080-2d6f-11e9-8e9e-9118acd2e924.jpg)

That's [Unicorn Calligraphy](https://www.fontspace.com/mistis-fonts/unicorn-calligraphy)

**Comment by Frenzie** [comment]
@NiLuJe 

> I'm not quite sure what's at play here, but I'd say some fairly aggressive kerning, I can confirm the underlying cause w/ FBink too.

That rendering of `fl` is something I'd noticed prior to HB. (I find it much more pleasant to read EPUB/FB2 now in general.) But I only just realized it also happens when turning kerning off.

**Comment by poire-z** [comment]
> Although, taking an actual look at the font, that's the LSB & RSB, the glyph metrics should provide 'em. They just happen to be negative in these cases, which the code might not be expecting?

I guess it's fine (and actually expected) for the code to not deal with that, as it's for writting a bit of this glyph over the next glyph.
crengine does just need to account for them at end of line (and start of line), which it does, but just when the font is italic.
It does it thanks to over metrics: `RSB = glyph.originX - glyph.width + glyph.blackBoxX`,` LSB= -glyph.originX` which is an alternative way of getting the right and left side bearings (or their positive values if there are negative).

**Comment by houqp** [comment]
calculating RSB for last character on each line adding 10+ seconds rendering time seems way more than I would expect. Is majority of the time spent in calling `font->getGlyphInfo(`?

**Comment by ptrm** [comment]
There is also case with [Libertinus Serif 6.6](https://github.com/libertinus-fonts/libertinus) to happen with "J" on the left edge (and full J for reference):
![smartselect_20190211-085952_gallery](https://user-images.githubusercontent.com/366258/52551114-6d848580-2ddb-11e9-97fc-16a62c13f6ed.jpg)

![smartselect_20190211-085429_gallery](https://user-images.githubusercontent.com/366258/52551041-24cccc80-2ddb-11e9-8500-9460ade20193.jpg)

**Comment by poire-z** [comment]
> calculating RSB for last character on each line adding 10+ seconds rendering time seems way more than I would expect. Is majority of the time spent in calling font->getGlyphInfo()?

Yes, it is quite expensive when done thousand of times. Even if we do it only for each line start and end (so, 2 times per line), it has that big impact (so, I can't imagine the time needed to render a book if we were not caching widths and call it for each char of a book ! :)
But caching it, just like we cache glyphs' widths, gets back my original rendering time. Working on it.

**Comment by ptrm** [comment]
Well, for the record, some italics also appear to be slightly cut, just encountered this yesterday ("_ll_"):
![smartselect_20190212-105528_gallery](https://user-images.githubusercontent.com/366258/52627115-fd4a3280-2eb4-11e9-9111-f604b1d697e9.jpg)

**Comment by poire-z** [comment]
(I've been fighting with fixing that with fake italic - for fonts that does not have italic - to no avail, but I think your Libertinus is not such a font.)
If I give you a libcrengine.so built for Kobo with my fixes up to now, are you willing to give it a try and see if/how much it solves (or not...) the things you noticed?
(I still see one pixel missing here and there ... can't find where I could grab some more... :)

**Comment by ptrm** [comment]
> If I give you a libcrengine.so built for Kobo with my fixes up to now, are you willing to give it a try and see if/how much it solves (or not...) the things you noticed?

@poire-z sure, I will be glad to give it a try

**Comment by poire-z** [comment]
Here you go (untested on a real kobo, so keep a backup of the previous files - included libkoreader-cre.so which might not be needed, but in the past, I felt they needed to match how they were built): 
[libcrengine-kobo-font-sides-fix1.zip](https://github.com/koreader/koreader/files/2855619/libcrengine-kobo-font-sides-fix1.zip)


Things that I have tweaked:
- Glyphs near the left and right page margin should no more be cut, wirh regular fonts or real italic fonts (I still have the feeling there might be sometimes 1px missing with small font size...)
- Highlights (after the highlight is saved, so, when re-displayed by koreader on the page) should include the full glyph. Might still be some pixels missing on the right side when the highlight spans multiple lines.

Things that I didn't manage to tweak, or that look too complicated:
- The above for fake italic glyphs with a font that has no italic (such as that Unicorn font).
- The native highlights by crengine (when you are currently selecting text) will still not include those glyphs sides (but the second after you hit higlight and the dialog dismissed, and the native highlight removed and replaced by a koreader highlight, you should see it grow to the correct glyph margins).

**Comment by ptrm** [comment]
True, italics still get a pixel or two off. That's Libertinus:
![smartselect_20190212-131023_gallery](https://user-images.githubusercontent.com/366258/52634545-c893a680-2ec7-11e9-9c70-85cba8cec45c.jpg)

That's Noto Serif at the size of 30 (200dpi override for a 300dpi screen)
![smartselect_20190212-130255_gallery](https://user-images.githubusercontent.com/366258/52634547-c92c3d00-2ec7-11e9-98d3-630b8ecebc44.jpg)

And regulars look great 😃 (tested on Clara HD and Forma):
![smartselect_20190212-130404_gallery](https://user-images.githubusercontent.com/366258/52634548-c92c3d00-2ec7-11e9-9700-3dda5f2ba1d9.jpg)
![smartselect_20190212-130445_gallery](https://user-images.githubusercontent.com/366258/52634549-c92c3d00-2ec7-11e9-84c8-a991a863e80b.jpg)

Selections for ending f look great. The _ll_ thing is hard to distinguish in highlights, but it's actually a semi-transparent pixel. Here at the font gamma set to 4, thought maybe that's the reason:
![reader_2019-feb-12_121709 crop](https://user-images.githubusercontent.com/366258/52635167-75baee80-2ec9-11e9-9c9a-ea9167757e84.png)
And here at gamma 0.8, maginified 2x:
![reader_2019-feb-12_121905 crop 2x](https://user-images.githubusercontent.com/366258/52635182-81a6b080-2ec9-11e9-8161-b0b250d8884c.png)
Still, I don't find it that annoying, it's just to serve you an edge case ;)

**Comment by ptrm** [comment]
And also confirmed to work with regular Unicorn font ;)

**Comment by poire-z** [comment]
Can you show what your glyph outlines software shows about the `W` uppercase and `l` lowercase, in regular and italic, for libertinus serif?
I get some extra leading space before the `W`, and the eating on the right of a `l` becomes bigger with a smaller font size (I added 1px to all last char on the right, enough with a large font, but not enough with a smaller font size):

Small font size (magnified x2 and x3 on the right):
![image](https://user-images.githubusercontent.com/24273478/52637297-24156280-2ecf-11e9-8ae6-e9df620de7cb.png) ![image](https://user-images.githubusercontent.com/24273478/52637361-545d0100-2ecf-11e9-9fc4-451987baf8c9.png)

Larger font size:

![image](https://user-images.githubusercontent.com/24273478/52637553-ce8d8580-2ecf-11e9-8616-7b06b32b0fa9.png) ![image](https://user-images.githubusercontent.com/24273478/52637596-f8df4300-2ecf-11e9-92a3-c4c2b093b3ce.png)

**Comment by ptrm** [comment]
> Can you show what your glyph outlines software shows about the `W` uppercase and `l` lowercase, in regular and italic, for libertinus serif?

It's [FontForge](http://fontforge.github.io/), btw.
Here you go (could not find the language switch on windows ;) ):
![image](https://user-images.githubusercontent.com/366258/52639264-262df000-2ed4-11e9-841b-610442e1fe40.png)
![image](https://user-images.githubusercontent.com/366258/52639354-57a6bb80-2ed4-11e9-8da3-7f3b8bc0aba1.png)

**Comment by ptrm** [comment]
Regarding fontforge, it has also its own scripting language and is availble to run headlessly.

There is also a useful tool for analysing letterspacing with kerning and other features (here for Baskervaldx, as Libertinus does not appear to have a kerning table):
![image](https://user-images.githubusercontent.com/366258/52640152-1dd6b480-2ed6-11e9-8fba-88d416d917fd.png)
![image](https://user-images.githubusercontent.com/366258/52640228-4ced2600-2ed6-11e9-854a-d9a06840d300.png)

**Comment by NiLuJe** [comment]
Here be dragons, because I have no idea how CRe/FT does this:

I fixed a similar issues in FBInk yesterday, and the gist of it turned out to be that when the final character of a line is a space (which was all of them once linebreaking kicked in), we don't print the space, but we *do* re-position the cursor and clip the bounding box based on that position.
The issue is that the cursor is positioned via the previous glyph's advance, which will often, by design, position you earlier than the RSB (or the edge of the glyph's *visible* bitmap bounding box).
Which means the cursor was actually going backwards, clipping the bounding box, and the edge of the previous glyph.
I had to special-case that for EoL spaces, honoring the previous's glyph's full width in case the cursor would be going backwards.

**Comment by poire-z** [comment]
crengine does it correctly the way you describe (but it did it only when the font was italic - it will now do it for any font).
It adds the abs(RSB) (when negative, so when it exceeds the advance) to the width of the last non space char in a line (when line breaking), so the glyph is said to be bigger than it is, so pushed a little bit on its left when drawing, so no overflow on the right margin.
https://github.com/koreader/crengine/blob/66166c0d6b823ec870a1a68b55181a5cfeff95e7/crengine/src/lvtextfm.cpp#L1656-L1660
https://github.com/koreader/crengine/blob/66166c0d6b823ec870a1a68b55181a5cfeff95e7/crengine/src/lvtextfm.cpp#L644-L660

So, the simple Freetype recipe, which seems alright, is:
`int delta = glyph.originX - glyph.width + glyph.blackBoxX;`

I'll be adding this in getGlyphInfo:

```c
         updateTransform();
         int error = FT_Load_Glyph(
             _face,          /* handle to face object */
             glyph_index,   /* glyph index           */
             flags );  /* load flags, see below */
         if ( error )
             return false;
         glyph->blackBoxX = (lUInt8)(_slot->metrics.width >> 6);
         glyph->blackBoxY = (lUInt8)(_slot->metrics.height >> 6);
         glyph->originX =   (lInt8)(_slot->metrics.horiBearingX >> 6);
         glyph->originY =   (lInt8)(_slot->metrics.horiBearingY >> 6);
         glyph->width =     (lUInt8)(myabs(_slot->metrics.horiAdvance) >> 6);
+        glyph->rsb =       (lInt8)( (myabs(_slot->metrics.horiAdvance)
+                            - _slot->metrics.horiBearingX - _slot->metrics.width) >> 6);
+                            // This >>6 on a negative number will floor() it, so we'll
+                            // get a ceil()'ed right overflow, which is what we want.
+        printf("%c %d: orX: %d + bbX: %d - adv: %d = rsb: %d (>>6 = %d)\n", code, _italic,
+            _slot->metrics.horiBearingX, _slot->metrics.width, _slot->metrics.horiAdvance,
+            _slot->metrics.horiBearingX + _slot->metrics.width - myabs(_slot->metrics.horiAdvance),
+            glyph->rsb);
+
+//        glyph->rsb -= 1;
+        // Note: when the font does not provide italic glyphs (_italic = 2), some fake
+        // italic/oblique is obtained with FreeType transformation (_matrix.xy and
+        // FT_Set_Transform()). freetype.h states about it:
+        //     Note that this also transforms the `face.glyph.advance' field,
+        //     but *not* the values in `face.glyph.metrics'.
+        // So, with such fake italic, the values just computed above are wrong,
+        // and may cause some glyphs to be truncated in the page right margin.
+        // Some possible attempt at guessing the transformed values can be found in
+        // http://code.qt.io/cgit/qt/qtbase.git/tree/src/platformsupport/fontdatabases/freetype/qfontengine_ft.cpp
+        // (transformBoundingBox) but a straightforward port here does not give
+        // the expected rendering...
```

Which would give with the above screenshoted `W` and `l`
Regular:
```
W 0: orX: 0 + bbX: 1408 - adv: 1408 = rsb: 0 (>>6 = 0)
l 0: orX: 0 + bbX: 384 - adv: 384 = rsb: 0 (>>6 = 0)
```
Italic:
```
W 1: orX: 192 + bbX: 1344 - adv: 1344 = rsb: 192 (>>6 = -3)  will increase the width by 3px
l 1: orX: 128 + bbX: 384 - adv: 384 = rsb: 128 (>>6 = -2) will increase the width by 2px
```
which seem a bit different from what I could guess from the drawings (on the pictures, the `l` don't have the same blackbox - I get 384 for both - and the italic W seem smaller (92% on the picture) than the 95% I get from freetype)... Dunno if that's a problem...

(I guess then that we should also account for the positive originX 3px for italic `W` and 2px for italic `l` when it's the first char of a line, but that's more complicated :)

Anyway, playing around again with fonts, I don't seem to be able to reproduce what I screenshoted above!

May be I was using a libertinusserifdisplay-regular.otf I had around (which I have no -italic.otf for, so I was in my fake italic case).
Using libertinusserif-regular.otf (which has a libertinusserif-italic.otf buddy), I don't get any of the cut on the right.

@ptrm : can you check if the fonts you have problems with do have a -italic buddy file, or not?

And if yes you have them :( : is it different depending on contrast gamma (best set to 1), with/without hinting/kerning (best without, so we don't have them in the process), in case that's these that make the bleeding...

**Comment by poire-z** [comment]
For easier checking, my test file :) [test-leftrightchar.html.txt](https://github.com/koreader/koreader/files/2856938/test-leftrightchar.html.txt)

**Comment by Frenzie** [comment]
This is probably completely unrelated, but just bringing it up since we're talking about spacing issues: sometimes there's no spacing after a split word, like so:

```
spit-
tingnext word
```

When it should be like

```
spit-
ting next word
```

I vaguely seem to recall this having been mentioned somewhere, but I figured I'd bring it up. I can try to create a test case tomorrow or the day after if necessary.

**Comment by ptrm** [comment]
@poire-z judging from their shape I do have separate serif files, "l"s differ to much to be just slanted ;)

Now ecerything looks great in all cases, hard to say what's going on :D Here are the screenshots and font listing: https://gist.github.com/2834777b90cfe36a0e3039ea635fa4b6

**Comment by poire-z** [comment]
> Now everything looks great in all cases, hard to say what's going on :D

You mean on my test html file only?
Or including with the books you used for your earlier screenshots? (which would mean both of us have moved in another dimension...)

(You have like me a single libertinusserifdisplay-regular.otf, so you should see bad and cut italic with it - as well as with FreeSerif which comes too without italic in koreader.)

So, to be clear: we should forget our early afternoon tests and consider all is fine (with the above fake italic limitations) ? :)

> I can try to create a test case tomorrow or the day after if necessary.

@Frenzie : not that fun to debug and fix, as I thought I was done with these kind of bugs... but please do.

**Comment by ptrm** [comment]
@poire-z I will double check my ebooks, but mind you: there's `libertinusserif-{regular,italic,...}` and `libertinusserifdisplay-regular` ;)

**Comment by Frenzie** [comment]
@poire-z It seems to be a problem specifically with the way I like to read, namely with full-on HarfBuzz ("best").

![screenshot_2019-02-12_20-49-53](https://user-images.githubusercontent.com/202757/52663879-f993cb80-2f07-11e9-89c7-772acc220a63.png)

[diffuse.html.zip](https://github.com/koreader/koreader/files/2857419/diffuse.html.zip)

**Comment by ptrm** [comment]
And there indeed are pixels dangling outside the highlight (not cre selection) for Libertinus Serif (without Display for sure ;) ) italic (note the l's bottoms). Gammas are 1 and 4, hinting off, kerning best:
![smartselect_20190212-204817_fx](https://user-images.githubusercontent.com/366258/52664023-4c6d8300-2f08-11e9-9db1-6d96660670c6.jpg)
![smartselect_20190212-204909_fx](https://user-images.githubusercontent.com/366258/52664037-53949100-2f08-11e9-8dcd-482587dee666.jpg)

**Comment by Frenzie** [comment]
@poire-z On my H2O it actually looked like this:

![fusecompassion](https://user-images.githubusercontent.com/202757/52664171-9eaea400-2f08-11e9-811b-a5c47dfb4f35.png)

Playing about a bit with the window size makes even worse results possible (e.g., the very last one)
![screenshot_2019-02-12_20-56-36](https://user-images.githubusercontent.com/202757/52664227-c4d44400-2f08-11e9-87f9-bfc9024213f6.png)

**Comment by ptrm** [comment]
So, it appears italics lean out of bounds on kerning turned on personally I really don't mind the half-transparent pixel outside, the change after the fix posted above is huge for me anyway ;)

**Comment by ptrm** [comment]
Closing was an accident ;)

**Comment by Frenzie** [comment]
Pinging @virxkane

Have you already noticed https://github.com/koreader/koreader/issues/4577#issuecomment-462908911 and https://github.com/koreader/koreader/issues/4577#issuecomment-462910716?

**Comment by poire-z** [comment]
> So, it appears italics lean out of bounds on kerning turned on

So, everything's ok when kerning is off?
What about when it is "fast"? and "good"?

If it happens only with best or good, it might be something else to look at that what I focused on :)

@Frenzie : only with "best", not with "good" ?

Curious to hear about how it does on coolreader.
It's quite possible it's due to some other text leading/trailing spacing fix of mine, and that it would not happen on coolreader...

**Comment by poire-z** [comment]
And please test with gamma 1. I guess any gamma > 1 enlarge the pixels, so may cause some bleeding. And there's not much we can do to know the amount of that and account for it.

**Comment by Frenzie** [comment]
@poire-z Indeed, the effect doesn't occur in good. In fact the spacing is altogether different. Mostly smaller and for the better I'd say.

**Comment by ptrm** [comment]
>So, everything's ok when kerning is off?
>What about when it is "fast"? and "good"?
>
> If it happens only with best or good, it might be something else to look at that what I focused on :)

@poire-z It bleeds on the best kerning setting, also with gamma 1 (one of two screenshots in my second last commit is the gamma 1 / kerning best one), I will do a more comprehensive testing tomorrow and cover all kerning settings.

**Comment by poire-z** [comment]
To make the issue wider:
We were talking above about negative RSB: glyph overflowing its advance box.

As seen on the following screenshots, there can be glyphs at start of line that have a positive LSB, or at end of line with a positive RSB. These positive SBs add some spacing, which make the glyph a bit indented, so possibly misaligned with some other glyphs on other lines on that page margin.

![image](https://user-images.githubusercontent.com/24273478/52637553-ce8d8580-2ecf-11e9-8616-7b06b32b0fa9.png)  ![image](https://user-images.githubusercontent.com/24273478/52637361-545d0100-2ecf-11e9-9fc4-451987baf8c9.png)

Should we account for these, and so shift the glyphs/reduce its width, so not only dealing with overflows like I named my caches and functions, but also with underflows?
Or can we consider these underflows as some font characteristics, that do make sense in the middle of text, but could also make sense to keep the spacing at start or end of lines?
(I feel like opening a can of worms if I begin to play with all that - as we never really ever noticed it was a problem :) except for the more obvious overflows and truncated glyphs...)

**Comment by ptrm** [comment]
As for me, If the glyphs weren't truncated atthe end of lines, as floating punctuation isn't, I would be equally glad to have it. Unless my imagination fails ;)

**Comment by poire-z** [comment]
Unrelated, but the glyph widths cache (and glyphinfo structs) use a lUint8. So it works with glyphs width up to 254 px (0xFF is reserved to mean unitiliazed).
This means rendering glitches with super huge font sizes as a width of 290px is stored and get back as 36px or something. I could at minima ensure it is stored as 254.
Or shall we really care about super huge fonts, and use lInt16 and lUint16 to store and cache metrics... ?

**Comment by ptrm** [comment]
@poire-z So, the issue with excess pixels in selection and screen edges is dependant on font size, not kerning. Screenshots are lengthy, so here is the gist with the description and images: 

https://gist.github.com/ptrm/aac1dd78af69545d541b62918683bfd0
>Issue seems to be present at font size 22 and 44 (300dpi overriden in the menu to 200dpi, if that matters for the measures), regardless of kerning settings. Hinting is off.
For 24, though, there are no excess pixels.

Hope that 22 is small enough on 300dpi not to fall for the abovementioned lUint8 issue ;)

**Comment by poire-z** [comment]
I hardly notice anything on your screenshots :) I have to zoom a lot to see the semi transparent pixel, otherwise, it seems to adjust to the "blackbox". Dunno if the antialising (which happens even when no hinting) is responsible for the excess, or if it is supposed to always be accounted in the glyph metrics, or if it is allowed to bleed over.
I'm so going to assume that's just antialiasing you notice :)

Just to be sure, about the possible glyph cut (or bleeding/antialisaing cut) on the right margin (I guess it's acceptable in the text selection): do you notice it the same in the page with a blank margin, and in the area with the yellow border? Because there is another thing at play in the blank page margin (and not when inside the yellow border): the page cropping, which I don't suspect to be the cause of all that, but I just want to be sure.

(You should not notice any cut on the page margins when switching to scroll mode, because there is no such cropping, so the bleeding has available room to spread.)

**Comment by ptrm** [comment]
> I hardly notice anything on your screenshots :) I have to zoom a lot to see the semi transparent pixel, otherwise, it seems to adjust to the "blackbox".
@poire-z oh, I had non-antialiased zoom at hand which pixelated things for me :D The pixel is there and it's not any artifact which was added after screenshot was taken:
![smartselect_20190213-134134_fx](https://user-images.githubusercontent.com/366258/52712539-322db680-2f95-11e9-9bfa-fdf97362c9f5.jpg)

So, my question is, did you mean CRE antialiasing or screenshot viewer antialiasing?

As for the right margin cut, true, scroll mode also causes those excess pixels to disappear, as does the yellow border:
![smartselect_20190213-134825_fx](https://user-images.githubusercontent.com/366258/52712947-60f85c80-2f96-11e9-904b-4370d9b9e908.jpg)

**Comment by poire-z** [comment]
I meant CRE antialising, or even Freetype antialising.
Or some rounding error somewhere...
The freetype metrics we get are all multiple of 64, so after >>6, they get to 1px.
So, may be it's some rounding error in freetype. There are options to load/render/do subpixels rendering, which I don't really want to discover :)

> But I emphasize, that's just to provide you with corner cases,

I know, thanks :) I'm just having quick guesses and tests, to see if there's an easy fix or bug, that would happen to be in the area I understand :) If not, well, I won't dig too much.

**Comment by ptrm** [comment]
For reference, here is what Plato does to Libertinus Serif at the ends and beginnings of lines, and what for me is even more desired for me:

Left | Right
------------ | -------------
![image](https://user-images.githubusercontent.com/366258/52717477-7aeb6c80-2fa1-11e9-96f7-c94660f921ad.png) | ![image](https://user-images.githubusercontent.com/366258/52717029-6bb7ef00-2fa0-11e9-92e8-d656d1e159ba.png)

But then again, if the cropping doesn't come from margins, I won't push you much to do it this way ;)

**Comment by poire-z** [comment]
I guess that's what we would have if we do nothing else than just drop the clipping then.

Curious to see, if Plato supports border, what it would do with:
`<div style="border: 20px solid red; margin: 0; padding: 0">add your text here</div>.`
Would the extra glyph parts be rendered over/under the border, or does it account somehow for them in that case?

**Comment by ptrm** [comment]
> and how plato does with our italic _Will_ and _Jogof_: does it stand out in the margin? Is the effect also preferable when things go over in the margin (when surrounded by regular text lines that don't stand out)?

It doesn't ;) I took the time to create an epub out of the html attached above, as Plato doesn't support htmls it seems, and the result is no border in sight. Elements within the bordered box have the same absolute distance from the left screen edge as the ones outsie of the box ;)

Here is the epub: [test-leftrightchar.epub.gz](https://github.com/koreader/koreader/files/2860830/test-leftrightchar.epub.gz). It has Libertinus regular and italic embedded for your convenience ;)

Microsoft Edge, if its of any interest as I've had it at hand, lets characters overlap the border: 
![image](https://user-images.githubusercontent.com/366258/52721114-47144500-2fa9-11e9-9556-98b4b9056787.png) ![image](https://user-images.githubusercontent.com/366258/52720758-8aba7f00-2fa8-11e9-8380-2a276c1c8c84.png) ![image](https://user-images.githubusercontent.com/366258/52720788-973ed780-2fa8-11e9-9948-659934a5d8ea.png)

Calibre does this as well:
![image](https://user-images.githubusercontent.com/366258/52721056-1f24e180-2fa9-11e9-83d5-2468172c2e33.png)

**Comment by ptrm** [comment]
Chrome behaves in the same way as pictured above. Adding `overflow: hidden;` on the bordered div, though, clips the characters.

**Comment by poire-z** [comment]
Well, quick try at having crengine do that:
- no clipping
- draw the border before drawing the text (it currently draws border after the box content, so possibly over it).

```c
--- a/crengine/src/lvdocview.cpp
+++ b/crengine/src/lvdocview.cpp
@@ -1884,6 +1884,7 @@ void LVDocView::drawPageTo(LVDrawBuf * drawbuf, LVRendPageInfo & page,
                //clip.top = info.bottom;
        }
        drawbuf->SetClipRect(&clip);
+        drawbuf->SetClipRect(NULL);
        if (m_doc) {
                if (page.type == PAGE_TYPE_COVER) {
                        lvRect rc = *pageRect;
@@ -2236,6 +2237,7 @@ void LVDocView::Draw(LVDrawBuf & drawbuf, int position, int page, bool rotate, b
                return;
        if (isScrollMode()) {
                drawbuf.SetClipRect(NULL);
+        drawbuf.SetClipRect(NULL);
         drawPageBackground(drawbuf, 0, position);
         int cover_height = 0;
                if (m_pages.length() > 0 && m_pages[0]->type == PAGE_TYPE_COVER)
--- a/crengine/src/lvrend.cpp
+++ b/crengine/src/lvrend.cpp
@@ -3951,6 +3951,8 @@ void DrawDocument( LVDrawBuf & drawbuf, ldomNode * enode, int x0, int y0, int dx
                     }
                 }

+                if ( !isTableRowLike )
+                    DrawBorder(enode,drawbuf,x0,y0,doc_x,doc_y,fmt);
                 for (int i=0; i<cnt; i++)
                 {
                     ldomNode * child = enode->getChildNode( i );
@@ -3978,8 +3980,8 @@ void DrawDocument( LVDrawBuf & drawbuf, ldomNode * enode, int x0, int y0, int dx
                 // Don't draw border for TR TBODY... as their borders are never directly
                 // rendered by Firefox (they are rendered only when border-collapse, when
                 // they did collapse to the cell, and made out the cell border)
-                if ( !isTableRowLike )
-                    DrawBorder(enode,drawbuf,x0,y0,doc_x,doc_y,fmt);
+//                if ( !isTableRowLike )
+//                    DrawBorder(enode,drawbuf,x0,y0,doc_x,doc_y,fmt);
                }
             break;
         case erm_list_item: // obsolete rendering method (used only when gDOMVersionRequested < 20180524)
@@ -4008,6 +4010,7 @@ void DrawDocument( LVDrawBuf & drawbuf, ldomNode * enode, int x0, int y0, int dx
                         txform->Draw( &drawbuf, doc_x+x0 + padding_left - list_marker_width, doc_y+y0 + padding_top, NULL, NULL );
                     }
                 }
+                DrawBorder(enode, drawbuf, x0, y0, doc_x, doc_y, fmt);

                 // draw whole node content as single formatted object
                 LFormattedTextRef txform;
@@ -4041,7 +4044,7 @@ void DrawDocument( LVDrawBuf & drawbuf, ldomNode * enode, int x0, int y0, int dx
                 drawbuf.FillRect( doc_x+x0+fmt.getWidth()-1, doc_y+y0, doc_x+x0+fmt.getWidth(), doc_y+y0+fmt.getHeight(), color );
                 drawbuf.FillRect( doc_x+x0, doc_y+y0+fmt.getHeight()-1, doc_x+x0+fmt.getWidth(), doc_y+y0+fmt.getHeight(), color );
 #endif
-                DrawBorder(enode, drawbuf, x0, y0, doc_x, doc_y, fmt);
+//                DrawBorder(enode, drawbuf, x0, y0, doc_x, doc_y, fmt);
                 /*lUInt32 tableBorderColor = 0x555555;
                 lUInt32 tableBorderColorDark = 0xAAAAAA;
                 bool needBorder = enode->getStyle()->display==css_d_table_cell;
```

<kbd>![image](https://user-images.githubusercontent.com/24273478/52722108-47addb00-2fab-11e9-9634-e696fa0ba787.png)</kbd>
<kbd>![image](https://user-images.githubusercontent.com/24273478/52722162-5e543200-2fab-11e9-94e9-a6e6ef987579.png)</kbd>

Don't know much about side effects, but we could have that same behaviour.
~~I'm just not sure I like it~~ I'm just not sure if it is preferable, but I'm sure I don't like it :)
We should look at our printed books for these italic J and f, to see how publishers/publishing tools do them:)

Or we can just keep the RSB LSB shifts fixes, get rid of the clipping, draw the border early (could be it's preferable to do it late), so you'd get your extra pixels.

(How do Edge/Calibre/Plato do with the huge leading left space at start of line with the italic _Will_? Kept, or corrected?)

**Comment by ptrm** [comment]
> I'm just not sure I like it I'm just not sure if it is preferable, but I'm sure I don't like it :)

My question is: Is it achievable to simply toggle the two on any level (defaults.lua or in the UI)? KOReader seems already to allow much liberty to users' typographic preferences ;)

> (How do Edge/Calibre/Plato do with the huge leading left space at start of line with the italic _Will_? Kept, or corrected?)

Exactly the same way as pictured above. Large space on the left. Noone said Libertinus is state-of-the-art :>

**Comment by poire-z** [comment]
> Is it achievable to simply toggle the two on any level

It's possible (but not fun to do, and I challenge you to find a toggle name and values, and an explanation for it :), but we're not there yet.
We would first need to find out what's the best, most common, most appreciated, as we would have to select a default anyway. (And I might probably not mind the one you prefer, as I usually read with a small font size and I won't notice it)

But... 2 things just occured to me:
What you prefer is actually the current crengine behaviour :)
And you could just have solved your original issue with a style tweak :) :
`body {padding-left: 0.2em; padding-right: 0.2em}` might just solve your cut glyphs issues, including with fake italic...
We could somehow enforce that internally without CSS if it's somehow the right solution.

Now, what about all the cans we opened in the current issue? Just close them and forget? :)

**Comment by poire-z** [comment]
@ptrm : here's a [libcrengine.zip](https://github.com/koreader/koreader/files/2864122/libcrengine.zip) with just horizontal clipping removed, which should solve your first post issue. It's just this:
```diff
--- a/crengine/src/lvdocview.cpp
+++ b/crengine/src/lvdocview.cpp
@@ -1853,11 +1853,15 @@ void LVDocView::drawPageTo(LVDrawBuf * drawbuf, LVRendPageInfo & page,
        //    offset = 0;
        int offset = 0;
        lvRect clip;
-       clip.left = pageRect->left + m_pageMargins.left;
        clip.top = pageRect->top + m_pageMargins.top + headerHeight + offset;
-       clip.bottom = pageRect->top + m_pageMargins.top + height + headerHeight
-                       + offset;
-       clip.right = pageRect->left + pageRect->width() - m_pageMargins.right;
+       clip.bottom = pageRect->top + m_pageMargins.top + height + headerHeight + offset;
+       // clip.left = pageRect->left + m_pageMargins.left;
+       // clip.right = pageRect->left + pageRect->width() - m_pageMargins.right;
+       // We don't really need to enforce left and right page margins: this
+       // allows glyphs that need to (like 'J' at start of line or 'f' at
+       // end of line with some fonts to not be cut by this clipping).
+       clip.left = pageRect->left;
+       clip.right = pageRect->left + pageRect->width();
        if (page.type == PAGE_TYPE_COVER)
                clip.top = pageRect->top + m_pageMargins.top;
     if (((m_pageHeaderInfo || !m_pageHeaderOverride.empty()) && page.type
```

Summary of other things noticed along this issue:
- border are drawn after their inner content, so they could override it if is bleeds a bit. In practice, their is always some padding that would avoid that, and black on black bleeding would not be noticed. Should the border drawing be done earlier, before the content? Any foreseen side effect I don't see?
- glyphinfo and width cache should use lInt16 instead of lInt8 for correct rendering of big font sizes
- Absence of cache for RSB can make books with big italic sections slow to load. Should these be cached?
- The highlighting of text selection currently does not include the negative LSB/RSB. Does this need fixing?

Regarding typographic usage of the above choice (have a `J` bleed in the left margin, or have it indentend so it does not bleed), I looked at some printed books, and in recent books (made with computers), it bleeds - on older (1950, 1970), it doesn't (probably due to hardware typographic material, metalic glyphs than can't bleed in a metalic margin :). I still don't know what's prefered among typographic aesthetes.
(In that recent book, italic text bleeds like hell in the right margin, it's really ugly - so, crengine actually did right in counteracting that for italic.)

Regarding @Frenzie rendering issue with Harfbuzz full, I think it's because of the ligature. Text fragments are measured only once, so `diffuse` is measured with the `ff` ligatured. Later, hyphenation decides a cut at `dif-fuse` is allowed, and when it's used, the measurement is wrong because it's no more ligatured, and one of the 2 `f` is of width 0, and the other one carries the width of the ligatured `ff`.
I noticed no rendering problem if I substitude `diffuse` with `dirfuse`.
Not sure how to go about that. Drop the allowed hyphenation flag when it happens to be on a ligature?

**Comment by ptrm** [comment]
> Regarding typographic usage of the above choice (have a `J` bleed in the left margin, or have it indentend so it does not bleed), I looked at some printed books, and in recent books (made with computers), it bleeds - on older (1950, 1970), it doesn't (probably due to hardware typographic material, metalic glyphs than can't bleed in a metalic margin :). I still don't know what's prefered among typographic aesthetes.
(In that recent book, italic text bleeds like hell in the right margin, it's really ugly - so, crengine actually did right in counteracting that for italic.)

True. I have a Polish copy of [Jost Hochuli's _Detail in Typography_](https://www.amazon.co.uk/Detail-Typography-Jost-Hochuli/dp/0907259340), and here is as much as I can publish without infringing any copyrights: https://www.dropbox.com/sh/b96ofdg0b8op5yi/AABzv9vJaaka1UByObMZUvfOa?dl=0
But then compare it with the back cover photo on Amazon: the punctuation does not bleed in the English version, while it does in Polish ;) As far as I could skim the book, there are so many rules and preferences and fashions, that I don't think there could be one answer, _especially_ a single one for serif, slab serif,  and sans serif fonts.

I post it because d2d publishing house is highly esteemed in the editing world in Poland, as far as a layman like me can tell, and I assumed that not only the contents, but the typesetting too are exemplary ;)

Also, I didn't really want to read the book too much, because the rules for applying ligatures for different languages, or maintaining line spacing while using quotes with different font sizes, might make me want to drop ereaders altogether :D

>(And I might probably not mind the one you prefer, as I usually read with a small font size and I won't notice it)
I too read with relatively small font sizes (so that on Aura One and Forma the page count is usually smaller than the published paperback, also with small type). But I also found it heavily depends on the language, as in Polish I could not find any example while reading (except for one quotation in English). In English I have at least one "of" at the end of the line every two pages.

And when it comes to styletweaking body paddings, or using floating typography, it's a great solution indeed. I would only prefer to have something maintaining the actual default.lua's margins without the need to calculate a total in memory ;)
So, in the end, the above fix sounds the best solution for me :)

**Comment by poire-z** [comment]
@Frenzie : regarding your Harfbuzz "best" issue:

"good" (no issue, as no ligature)
![image](https://user-images.githubusercontent.com/24273478/52787935-b51b4380-305f-11e9-9fbf-0b0dde2290d0.png)

"best" before (issues when a hyph wrap happens on a ligature):
![image](https://user-images.githubusercontent.com/24273478/52787993-e267f180-305f-11e9-90c8-93a668c4e82b.png)


"best" by just disallowing allowed hyphenation if it happens to be on a ligature:
![image](https://user-images.githubusercontent.com/24273478/52787902-a6cd2780-305f-11e9-8c23-c236afbafa6c.png)

Is that acceptable, even if it makes "best" a little less "best" ? :)
(I guess the correct way to fix this would be to drop/redo measurements of the whole/parts of the text involved, which looks very complicated to plug into the current code flow.)

**Comment by Frenzie** [comment]
@poire-z Looks kind of ugly to me compared to good, with the increase in word spacing due to a lack of hyphenation. On a non-artificial text the effect will hopefully be less pronounced.

But sure, if there's no apparent way to fix it in a reasonable amount of time just make sure there's sufficient TODO info in the comments. And I figure we might want to open a proper issue about it on crengine to not lose track of it when someone's looking for something to do.

**Comment by poire-z** [comment]
OK, some of these issues should be fixed by https://github.com/koreader/crengine/pull/261.
Spent a bit more of time on the HB full issue, and I found a better solution.

**Comment by ptrm** [comment]
Btw, I think I discovered some word spacing glitch with floating punctuation (much more spaces than when turned off, regardless of margins and word spacing settings I think), but that's for another issue ;> Just mentioning it in case it was something already known by any of you.

Anyway, the fix zsynced from nightlies looks stable for Kobos. Thanks a lot!

**Comment by poire-z** [comment]
Yeah, floating punctuations has its own set of glitches. See https://github.com/koreader/koreader/issues/2844.
As it's mostly important to CJK readers (and now, polish ones I discovered :), and there are lots of CJK specifics in the floating punctuation code, I'd rather not investigate/touch any of it for risk of breaking CJK stuff I won't be able to test/appreciate...

Although one of the point of this code ( [here](https://github.com/koreader/crengine/blob/fd290a9251d0de628794d716c4083098b838a546/crengine/src/lvtextfm.cpp#L1464-L1480) and [there](https://github.com/koreader/crengine/blob/fd290a9251d0de628794d716c4083098b838a546/crengine/src/lvtextfm.cpp#L1265-L1336) ) is to reduce the width of a line to make room for the punctuation/hyphens to be drawn outside of the text area, which was needed because of the clipping.
Now that we killed that clipping, it could be made easier to not just care about line width and draw punctuations/hyphens outside of the untouched regular line width...

**Comment by poire-z** [comment]
OK, corrected a few of the other issues in https://github.com/koreader/crengine/pull/263.

Highlights will now include the glyph overflows (and so, may overlap in some extreme rare cases).
May look ugly on our test page, but it's less ugly to have a gray line box stand out of a gray paragraph box, than a black glyph standiout of a gray parapgraph box.

<kbd>![image](https://user-images.githubusercontent.com/24273478/52913709-1857e080-32c1-11e9-8290-e8eb416a477b.png)</kbd>

Underflows are not corrected, because that would cause many indent/dedent in a whole selected paragraph rectangle, which would be really ugly and noticable.
